### DISCUSSION TITLE
Configurable Prisoner's Dilemma simulation model

### DISCUSSION THREAD

**Discussion by zeyus** [FLAMEGPU/FLAMEGPU2#899] [Show and tell]
We've been working on this for a couple of weeks for an exam project. 

## What is it?

A 2D ABM simulating interactions of "games" between agents, specifically the prisoner's dilemma game, in which participants can either cooperate or defect, which has a payoff matrix.

<img width="326" alt="image" src="https://user-images.githubusercontent.com/75656/184105191-1f7af765-add8-4161-9998-062c39f65c18.png">
Axelrod, R., & Hamilton, W. D. (1981). The Evolution of Cooperation. Science, 211(4489), 1390–1396. https://doi.org/10.1126/science.7466396

In this setup the default payoff matrix for interactions is

|   | They Cooperate | They Defect |
|:---------:|:------------:|:---------:|
| I Cooperate | 3.0 | -1.0 |
| I Defect | 5.0 | 0.0 |

Agents play all of their neighbours and if they cannot, they move, and pay a cost of travel.

After play or movement each agent has the opportunity to reproduce if they have sufficient energy and an available space around them, if they reproduce, they pay the energy cost of reproduction.

Each round there is a cost of living imposed on all agents except agents that were just born in the reproduction phase.

Any agents that drop below zero energy die.

## Other features

- Agent strategy distributions can be configured (more than 4 strategies broken at the moment, but strategy probability can be set to 0.0)
- Agents have a random trait assigned to them (default 1 of 4 possible traits) (more than 4 traits broken at the moment)
- Strategy mutation can be configured at a specific mutation rate which applies during reproduction
- Agents can employ a global strategy (i.e. always cooperate, with any agent) or a strategy for agents with the same trait (kin) or others, OR (reporting broken for this but it works) a strategy per unique trait)
- Environmental noise can be configured for a chance of miscommunication (i.e. if i choose cooperate, it becomes defect)
- Can run in a CUDAEnsemble for a whole suite of simulation runs
- logging is configured for both single and multi runs, currently it collects the agent counts by their strategies, but it should also not bother doing any counts (for performance) if logging is disabled, which it still does

It was fun, I learned a lot, and the code is part spaghetti, part anti-pattern, part polypattern.

I'm fully aware that some things could be way more optimized, compartmentalized and cleaner, there are a bunch of TODOs and I'll slowly get around to adding them to the github issues and chip away at them but any suggestions welcome.

Repo here: https://github.com/zeyus/FLAMEGPU2-Prisoners-Dilemma-ABM

<img width="1238" alt="Socult ABM model flow-General flow drawio" src="https://user-images.githubusercontent.com/75656/184108979-10fbb3d9-32f0-4610-9941-a67593097527.png">

![screenshot3_abm_paper_half_size](https://user-images.githubusercontent.com/75656/184108676-8f6821eb-f792-484c-b4a8-ba02de789a1f.png)

**Comment by Robadob** [comment]
It's great to see that you've achieved what you set out to do, and I hope you get a good mark for your work.

I may look further into this when I have more time (I haven't had a change to look at the paper you've referenced).

Thanks for sharing this.

**Reply by zeyus** [reply]
> It's great to see that you've achieved what you set out to do, and I hope you get a good mark for your work.

Thanks, could not have done it without your help! My friend and collaborator is seriously considering building upon this for his Bachelor's thesis, and while I haven't come close to deciding what I want to do, it's a tempting option to join him on that. A lot of things we implemented just added way too many dimensions so we didn't even end up using environmental noise, mutation or strategies per trait for the exam, so without a lot more work there are already plenty of options.

> I may look further into this when I have more time (I haven't had a change to look at the paper you've referenced).

That paper is one of the earlier investigations into the prisoner's dilemma and outlines the concepts, the main two papers we adapted from and expanded on were these two, which are more relevant to ABM simulations of PD style games:

- Smaldino, P., Schank, J., & Mcelreath, R. (2013). Increased Costs of Cooperation Help Cooperators in the Long Run. The American Naturalist. https://doi.org/10.1086/669615
- Hammond, R. A., & Axelrod, R. (2006). Evolution of contingent altruism when cooperation is expensive. Theoretical Population Biology, 69(3), 333–338. https://doi.org/10.1016/j.tpb.2005.12.002

I will also add what we wrote about the ABM (the exam) to the repo after exam grading is done (don't want to risk accidentally getting marked for cheating)

> Thanks for sharing this.

You're very welcome, I got out of industry and started my academic path for this very reason!

**Comment by MILeach** [comment]
Looks like really nice work, thanks for staying so communicative about your project!

**Reply by zeyus** [reply]
> Looks like really nice work, thanks for staying so communicative about your project!

You're welcome, it was a marathon...and it would be amazing if anyone else could get use out of the work or even some inspiration. (It could be a template of what not to do hahaha)

**Comment by mondus** [comment]
Just to add. Really nice work!

**Reply by zeyus** [reply]
Thank you! By the way that transpiler is looking amazing! Over the next couple of weeks I'm going to do a little exploration into making swig generate R bindings for FLAMEGPU2, it may be beyond me, but if not, I'll make it into a proper PR so all those data scientists can play around with it too. Of course, it would benefit from a transpiler too, but that's not something that can be ported, the syntax is so different to python.

**Reply by mondus** [reply]
That would be fantastic. I'm hoping to get the transpiler merged int his week once the docs are completed. Please let us know any progress you make with R!
